File: src/detect/python.ts
import { join } from 'node:path';
import { exists, readText } from '../utils/fs.js';

type PartialStack = Omit<import('./index.js').DetectedStack, 'features'>;

export async function detectPython(cwd: string): Promise<PartialStack | null> {
	const hasPyproject = await exists(join(cwd, 'pyproject.toml'));
	const hasRequirements = await exists(join(cwd, 'requirements.txt'));

	if (!hasPyproject && !hasRequirements) return null;

	const result: PartialStack = {
		language: 'python',
		framework: null,
		preset: null,
		testRunner: null,
		linter: null,
		typeChecker: null,
		formatter: null,
		packageManager: null,
	};

	// Read deps from available files
	let depsText = '';
	if (hasPyproject) {
		depsText = await readText(join(cwd, 'pyproject.toml'));
	}
	if (hasRequirements) {
		depsText += '\n' + await readText(join(cwd, 'requirements.txt'));
	}

	const hasDep = (name: string) => depsText.includes(name);

	// Framework
	if (hasDep('fastapi')) {
		result.framework = 'fastapi';
		result.preset = 'python-fastapi';
	} else if (hasDep('django')) {
		result.framework = 'django';
		result.preset = 'python-django';
	} else if (hasDep('flask')) {
		result.framework = 'flask';
		result.preset = 'python-flask';
	}

	// Test runner
	if (hasDep('pytest')) {
		result.testRunner = { name: 'pytest', command: 'pytest' };
	} else {
		result.testRunner = { name: 'unittest', command: 'python -m unittest discover' };
	}

	// Linter
	if (hasDep('ruff')) {
		result.linter = { name: 'ruff', command: 'ruff check .' };
	} else if (hasDep('flake8')) {
		result.linter = { name: 'flake8', command: 'flake8' };
	} else if (hasDep('pylint')) {
		result.linter = { name: 'pylint', command: 'pylint src/' };
	}

	// Type checker
	if (hasDep('mypy')) {
		result.typeChecker = { name: 'mypy', command: 'mypy .' };
	} else if (hasDep('pyright')) {
		result.typeChecker = { name: 'pyright', command: 'pyright' };
	}

	// Formatter
	if (hasDep('ruff')) {
		result.formatter = { name: 'ruff', command: 'ruff format .' };
	} else if (hasDep('black')) {
		result.formatter = { name: 'black', command: 'black .' };
	}

	// Package manager
	if (await exists(join(cwd, 'poetry.lock'))) {
		result.packageManager = 'poetry';
	} else if (await exists(join(cwd, 'pdm.lock'))) {
		result.packageManager = 'pdm';
	} else if (await exists(join(cwd, 'uv.lock'))) {
		result.packageManager = 'uv';
	} else {
		result.packageManager = 'pip';
	}

	return result;
}

File: internal/detect/python.go
package detect

import (
	"github.com/samahlstrom/forge-cli/internal/util"
	"path/filepath"
	"strings"
)

// DetectPython detects Python projects.
func DetectPython(cwd string) *DetectedStack {
	hasPyproject := util.Exists(filepath.Join(cwd, "pyproject.toml"))
	hasRequirements := util.Exists(filepath.Join(cwd, "requirements.txt"))
	if !hasPyproject && !hasRequirements {
		return nil
	}

	result := &DetectedStack{Language: "python"}

	var depsText string
	if hasPyproject {
		if t, err := util.ReadText(filepath.Join(cwd, "pyproject.toml")); err == nil {
			depsText = t
		}
	}
	if hasRequirements {
		if t, err := util.ReadText(filepath.Join(cwd, "requirements.txt")); err == nil {
			depsText += "\n" + t
		}
	}

	has := func(name string) bool { return strings.Contains(depsText, name) }

	// Framework
	switch {
	case has("fastapi"):
		result.Framework = "fastapi"
		result.Preset = "python-fastapi"
	case has("django"):
		result.Framework = "django"
		result.Preset = "python-django"
	case has("flask"):
		result.Framework = "flask"
		result.Preset = "python-flask"
	}

	// Test runner
	if has("pytest") {
		result.TestRunner = &ToolInfo{"pytest", "pytest"}
	} else {
		result.TestRunner = &ToolInfo{"unittest", "python -m unittest discover"}
	}

	// Linter
	switch {
	case has("ruff"):
		result.Linter = &ToolInfo{"ruff", "ruff check ."}
	case has("flake8"):
		result.Linter = &ToolInfo{"flake8", "flake8"}
	case has("pylint"):
		result.Linter = &ToolInfo{"pylint", "pylint src/"}
	}

	// Type checker
	switch {
	case has("mypy"):
		result.TypeChecker = &ToolInfo{"mypy", "mypy ."}
	case has("pyright"):
		result.TypeChecker = &ToolInfo{"pyright", "pyright"}
	}

	// Formatter
	switch {
	case has("ruff"):
		result.Formatter = &ToolInfo{"ruff", "ruff format ."}
	case has("black"):
		result.Formatter = &ToolInfo{"black", "black ."}
	}

	// Package manager
	switch {
	case util.Exists(filepath.Join(cwd, "poetry.lock")):
		result.PackageManager = "poetry"
	case util.Exists(filepath.Join(cwd, "pdm.lock")):
		result.PackageManager = "pdm"
	case util.Exists(filepath.Join(cwd, "uv.lock")):
		result.PackageManager = "uv"
	default:
		result.PackageManager = "pip"
	}

	return result
}
